### DISCUSSION TITLE
Failing to import a database with plpgsql functions

### DISCUSSION THREAD

**Discussion by scvalex** [cloudnative-pg/cloudnative-pg#969] [General]
TL;DR I'm trying to bootstrap import a database with `plpgsql` functions, but it fails with `pg_restore: error: could not execute query: ERROR:  language "plpgsql" does not exist`.  I'm probably not the first person to try this, so what's the trick to getting this to work?

Details:

I'm trying to convert an old PostgreSQL pod to cloudnative-pg.  I'm using the microservice import:

```
apiVersion: postgresql.cnpg.io/v1
kind: Cluster
metadata:
  name: pg-data
spec:
  instances: 2
  primaryUpdateStrategy: unsupervised
  imageName: ghcr.io/cloudnative-pg/postgresql:14.5
  storage:
    storageClass: "openebs-zfspv-lazy"
    size: 100Gi
  bootstrap:
    initdb:
      import:
        type: microservice
        databases:
          - dosare
        source:
          externalCluster: old-cluster
  externalClusters:
    - name: old-cluster
      connectionParameters:
        host: postgres
        user: dosare
        dbname: dosare
      password:
        name: dosare-config
        key: pg-password
```

This fails during the `pg_restore` with:

```
...
"level":"info","ts":1667911904.2648947,"logger":"pg_dump","msg":"pg_dump: dumping contents of table \"public.dosare\"","pipe":"stderr","logging_pod":"pg-data-1"}
{"level":"info","ts":1667911996.6173406,"msg":"dropping extensions from the target (empty) database","logging_pod":"pg-data-1"}
{"level":"info","ts":1667911996.6208682,"msg":"dropping extension from target database","logging_pod":"pg-data-1","extName":"plpgsql"}
{"level":"info","ts":1667911996.6550565,"msg":"temporarily granting superuser permission to owner user","logging_pod":"pg-data-1","owner":"app"}
{"level":"info","ts":1667911996.6687803,"msg":"executing database importing section","logging_pod":"pg-data-1","databaseName":"dosare","section":"pre-data"}
{"level":"info","ts":1667911996.6688378,"msg":"Running pg_restore","logging_pod":"pg-data-1","cmd":"pg_restore","options":["-U","postgres","--no-owner","--no-privileges","--role=app","-d","app","--section","pre-data","/var/lib/postgresql/data/pgdata/dumps/dosare.dump"]}
{"level":"info","ts":1667911996.7049005,"logger":"pg_restore","msg":"pg_restore: while PROCESSING TOC:","pipe":"stderr","logging_pod":"pg-data-1"}
{"level":"info","ts":1667911996.7049184,"logger":"pg_restore","msg":"pg_restore: from TOC entry 214; 1255 16391 FUNCTION diesel_manage_updated_at(regclass) dosare","pipe":"stderr","logging_pod":"pg-data-1"}
{"level":"info","ts":1667911996.7049224,"logger":"pg_restore","msg":"pg_restore: error: could not execute query: ERROR:  language \"plpgsql\" does not exist","pipe":"stderr","logging_pod":"pg-data-1"}
...
```

I tried to solve this a `postInitTemplateSQL`, but that doesn't seem to help:

```
      postInitTemplateSQL:
        - CREATE OR REPLACE LANGUAGE plpgsql;
```

I think what's going on is that the restore is dropping all the extensions, including `plpgsql` in https://github.com/cloudnative-pg/cloudnative-pg/blob/67285747a43754017e0a1fa16deafd00371aa1b7/pkg/management/postgres/logicalimport/database.go#L299

The comment there implies that `pg_restore` will recreate the extension, but that doesn't seem to be happening.  For reference, I also don't ever remember having to enable `plpgsql`; I always thought it was enabled by default.

Is there some way to tell cloudnative-pg to not drop extensions, or is there some other hook where I can add the extension during restore?

**Comment by scvalex** [comment]
Experimentally, a new database always has the `plpgsql` extension enabled:

```
$ pg_ctl initdb -D test-pg
...
$ pg_ctl start -D test-pg -w -o '-k /tmp'
...
$ psql -p 5432 -h localhost -d postgres
scvalex@postgres=# SELECT extname FROM pg_extension;
 extname
─────────
 plpgsql
(1 row)
```

We then run the almost the same `pg_dump` command that cloudnative-pg runs.  The only difference is that we use `-Fp` to output a plain SQL file instead of the custom `-Fc` format:

```
$ pg_dump -Fp -f test-pg.sql -d "host='localhost' user='scvalex' dbname='postgres'" -v
...
$ cat test-pg.dump
--
-- PostgreSQL database dump
--

-- Dumped from database version 14.5
-- Dumped by pg_dump version 14.5

-- Started on 2022-11-08 15:32:32 GMT

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET xmloption = content;
SET client_min_messages = warning;
SET row_security = off;

-- Completed on 2022-11-08 15:32:32 GMT

--
-- PostgreSQL database dump complete
--
```

There's no line in there to enable `plpgsql`.

Going by the [documentation](https://www.postgresql.org/docs/current/app-pgdump.html), when the `-e` option is not specified, "all non-system extensions in the target database will be dumped.".  I guess `plpgsql` is a system extension, so it doesn't get dumped.

**Comment by scvalex** [comment]
Reading [`pg_dump.c`](https://git.postgresql.org/gitweb/?p=postgresql.git;a=blob;f=src/bin/pg_dump/pg_dump.c;h=da427f4d4a17c177ca5658918125b7fc4204ba4b;hb=HEAD), it looks like there's an explicit filtering of "system extensions", where "system extension" is defined as an extension with an oid less than `#define FirstNormalObjectId 16384`.

You can actually circumvent this by deleting the `plpgsql` extension and re-creating it.  This causes it to have an oid that's larger than that hardcoded default, so `pg_dump` dumps it:

```
$ psql -p 5432 -h localhost -d postgres
scvalex@postgres=# drop extension plpgsql;
DROP EXTENSION
Time: 22.967 ms
scvalex@postgres=# create extension plpgsql;
CREATE EXTENSION
Time: 17.316 ms

$ pg_dump -Fp -f test-pg.sql -d "host='localhost' user='scvalex' dbname='postgres'" -v
...
$ cat test-pg.sql
--
-- PostgreSQL database dump
--

-- Dumped from database version 14.5
-- Dumped by pg_dump version 14.5

-- Started on 2022-11-08 15:47:21 GMT

SET statement_timeout = 0;
SET lock_timeout = 0;
SET idle_in_transaction_session_timeout = 0;
SET client_encoding = 'UTF8';
SET standard_conforming_strings = on;
SELECT pg_catalog.set_config('search_path', '', false);
SET check_function_bodies = false;
SET xmloption = content;
SET client_min_messages = warning;
SET row_security = off;

--
-- TOC entry 1 (class 3079 OID 16384)
-- Name: plpgsql; Type: EXTENSION; Schema: -; Owner: -
--

CREATE EXTENSION IF NOT EXISTS plpgsql WITH SCHEMA pg_catalog;


--
-- TOC entry 4316 (class 0 OID 0)
-- Dependencies: 1
-- Name: EXTENSION plpgsql; Type: COMMENT; Schema: -; Owner:
--

COMMENT ON EXTENSION plpgsql IS 'PL/pgSQL procedural language';


-- Completed on 2022-11-08 15:47:21 GMT

--
-- PostgreSQL database dump complete
--
```
```

**Comment by scvalex** [comment]
However, dropping and recreating the extension isn't a workaround because you can't drop the extension as long as there are functions using it:

```
dosare@dosare=# begin;
*dosare@dosare=# set constraints all deferred;
*dosare@dosare=# drop extension plpgsql;
ERROR:  cannot drop extension plpgsql because other objects depend on it
DETAIL:  function diesel_manage_updated_at(regclass) depends on language plpgsql
function diesel_set_updated_at() depends on language plpgsql
HINT:  Use DROP ... CASCADE to drop the dependent objects too.
Time: 34.336 ms
```

I don't see a viable workaround in either pg_dump or cloudnative-pg, so I think this is a bug in cloudnative-pg's `dropExtensionsFromDatabase`.  It incorrectly assumes that all extensions will be in the dump, but this is not the case for system extensions, most notably `plpgsql` which is how most people will have defined triggers.

**Comment by scvalex** [comment]
Created issue #970.

**Reply by gbartolini** [reply]
Thanks @scvalex - I have been reading through your investigation. It all resonates with me. I have produced a patch (#974), I would love if you could review it and approve it. Let me know if I can quote you in the commit message.

**Reply by scvalex** [reply]
The patch looks fine to me.  In particular, I have checked that the `oid >= 16384` is the right number and the right comparison operator.

No objection to being quoted on the commit.

**Reply by gbartolini** [reply]
Oh ... I think Jonathan missed this message and did not add you to the commit. We'll make sure to add you in the release notes.

**Reply by gbartolini** [reply]
The patch has been merged. You can either wait for 1.18.0 to be out or test the latest snapshot following the docs.

**Reply by scvalex** [reply]
Awesome.  Thanks!
